Fix race condition with async folder generation

AlanCoding opened the pull request:
##### SUMMARY
Proposed solution to https://github.com/ansible/ansible/issues/59306

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/utilities/logic/async_wrapper.py

##### ADDITIONAL INFORMATION
Unfortunately, due to the flaky nature of the error, I have only verified the fix manually, and doing so takes a good deal of time.

The method implemented in the safe mkdirs method seems to be a canned answer to the problem:

https://stackoverflow.com/a/273227/1092940

That's exactly how I would solve it. The docstring gives a lot of warnings about the appropriate place to use it, and I believe this case to fit that situation rather well.

Randomized files are used for anything actually _written_ to this directory. Even though the var name is `jobdir`, it refers to the global async directory, so the check for `EEXIST` is 100% appropriate.

Here are job runs inside of AWX, they start passing after this patch is applied:

![Screen Shot 2019-07-22 at 11 11 50 AM](https://user-images.githubusercontent.com/1385596/61643791-9144d500-ac71-11e9-834c-8dbafd292913.png)

AlanCoding committed:
Fix race condition with async folder generation

bcoca commented on lib/ansible/modules/utilities/logic/async_wrapper.py:
```diff
@@ -232,14 +233,13 @@ def _run_module(wrapped_cmd, jid, job_path):
-                "msg": "could not create: %s" % jobdir
-            }))
+    try:
+        makedirs_safe(jobdir)
```
fyi, this does not 'fix' race condition, it just makes it both less probable and ignore the error if it already exists

AlanCoding replied:
It's not an error if it already exists, because I _know_ that it already exists, and I know why. It's because it's ran locally against >1 host, so multiple processes are creating the same parent directory, but will create a randomized subdir. There is no issue with this, and it works just fine when the simultaneous processes do not hit the race condition.

bcoca commented on lib/ansible/modules/utilities/logic/async_wrapper.py:
```diff
@@ -232,14 +233,13 @@ def _run_module(wrapped_cmd, jid, job_path):
+    except Exception as e:
+        print(json.dumps({
+            "failed": 1,
+            "msg": "could not create %s: %s" % (jobdir, to_native(e))
```
you want to_text here as it is a json string, use to_native for when raising an exception

AlanCoding commented:
Can I get another look at this? We have been consistently hitting the same error since before this was opened.

AlanCoding committed:
Avoid use of safe_mkdirs import

Instead, manually add an except block to pass on the exists error
there is no reason to not pass on this error for the root async folder

AlanCoding commented:
bot_status

AlanCoding commented:
I've re-tested this in greater depth to be sure I have it right now.

playbook:

```yaml
---
- hosts: all
  gather_facts: false
  tasks:
  - name: Poll a sleep
    shell: "sleep 1"
    async: 30
    poll: 1
```

script:

```
total=100
n=0
e=0
while [ "$n" -lt $total ]; do
    n=$(( n + 1 ))
    rm -rf ~/.ansible_async
    ansible-playbook ~/Documents/repos/jlaska-ansible-playbooks/async_tasks.yml --connection=local -i tmp_hosts > /dev/null 2> /dev/null
    this_error="$?"
    # echo "this error $this_error"
    if [ $this_error == "99" ]; then
      echo "keyboard break"
      break
    elif [ $this_error == "2" ]; then
      echo "hit error case $n, code $this_error"
      e=$(( e + 1 ))
    else
      echo "success case $n, code $this_error"
    fi
done
echo "total runs $total"
echo "error runs $e"
```

If I run this with `devel`, it fails about 10 times out of 100.

Running with this branch now fails 0 times out of 100.

We are still hitting this error on a regular basis, and would really like to get this fix in so that we don't have to deal with these errors anymore.

AlanCoding commented:
and my new interpretation of @bcoca's prior comments is now that the safe mkdirs method wasn't really something that should have been imported. That's fine, I've tried to structure this to be much much more specific to the exact race condition involved in multiple local async processes.

AlanCoding commented:
ready_for_review

bcoca commented:
rebuild_merge

mkrizek reviewed (commented):
@AlanCoding can you add a changelog?

AlanCoding commented:
Closing in favor of https://github.com/ansible/ansible/pull/72069 which I like better too
